scripts/gb-model/compose_network.py

```diff
@@ -0,0 +1,311 @@
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class CompositionContext:
+    resources_root: Path
+    countries: tuple[str, ...]
+    costs_path: Path | None
+    costs_config: dict[str, Any]
+    max_hours: dict[str, Any] | None
+
+
+def create_context(network_path: str, config: dict[str, Any]) -> CompositionContext:
+    resources_root = Path(network_path).parents[1]
+    costs_cfg = copy.deepcopy(config.get("costs", {}))
+    cost_year = costs_cfg.get("year")
+    costs_path = (
+        resources_root / f"costs_{cost_year}.csv" if cost_year is not None else None
+    )
+
+    return CompositionContext(
+        resources_root=resources_root,
+        countries=tuple(config.get("countries", [])),
+        costs_path=costs_path,
+        costs_config=costs_cfg,
+        max_hours=config.get("electricity", {}).get("max_hours"),
+    )
+
+
+def _load_powerplants(
+    powerplants_path: Path | None,
+    costs,
+    config: dict[str, Any],
+) -> pd.DataFrame:
+    clustering_cfg = config.get("clustering", {})
+    consider_efficiency = clustering_cfg.get("consider_efficiency_classes", False)
+    aggregation_strategies = clustering_cfg.get("aggregation_strategies", {})
+    exclude_carriers = clustering_cfg.get("exclude_carriers", [])
+
+    columns = ["bus", "carrier", "p_nom", "max_hours"]
+
+    if powerplants_path is None:
+        logger.info(
+            "No power plant input provided; continuing without existing capacities"
+        )
+        return pd.DataFrame(columns=columns)
+
+    if not powerplants_path.exists():
+        logger.warning(
+            "Power plant data %s not found; continuing without existing capacities",
+            powerplants_path,
+        )
+        return pd.DataFrame(columns=columns)
+
+    return load_and_aggregate_powerplants(
+        str(powerplants_path),
+        costs,
+        consider_efficiency_classes=consider_efficiency,
+        aggregation_strategies=aggregation_strategies,
+        exclude_carriers=exclude_carriers,
+    )
+
+
+def integrate_renewables(
+    network: pypsa.Network,
+    config: dict[str, Any],
+    costs,
+    inputs: Mapping[str, str],
+    powerplants_path: Path | None,
+    hydro_capacities_path: Path | None,
+) -> None:
+    electricity_cfg = config.get("electricity", {})
+    renewable_carriers = list(electricity_cfg.get("renewable_carriers", []))
+
+    if not renewable_carriers:
+        logger.info("No renewable carriers configured; skipping integration")
+        return
+
+    if costs is None:
+        logger.warning("Cost data unavailable; skipping renewable integration")
+        return
+
+    extendable_carriers = copy.deepcopy(electricity_cfg.get("extendable_carriers", {}))
+    extendable_carriers.setdefault("Generator", [])
+
+    ppl = _load_powerplants(powerplants_path, costs, config)
+
+    profile_paths = {
+        key: Path(value) for key, value in inputs.items() if key.startswith("profile_")
```

Comment:
no need for the conditional if all input profiles are in a named snakemake input